From fa28f9ce6da8ff131b659885284839c98ebd8e87 Mon Sep 17 00:00:00 2001
From: Subbaraya Sundeep <sbhatta@marvell.com>
Date: Mon, 1 Apr 2019 19:27:51 +0530
Subject: [PATCH 128/386] octeontx2-pf: Remove flows when interface is down

Flows installed in MCAM need to be removed from ethtool
list otherwise flows disabled in hardware by NIX_LF_FREE
are displayed as active in ethtool.
Also ntuple filters deletion is allowed only when interface
is up which is same behavior when adding filters.

Change-Id: Id05f3196df4ec5bda5f359e3613767bdf4d24e85
Signed-off-by: Subbaraya Sundeep <sbhatta@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/6675
Tested-by: sa_ip-sw-jenkins
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[RH: Original patch taken from marvell 88F3720 board support SDK 10.0-PR2003]
Signed-off-by: Ruiqiang Hao <Ruiqiang.Hao@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c | 3 ++-
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c      | 1 +
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
index 6050773cb67e..50be9ae000b8 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
@@ -909,7 +909,8 @@ static int otx2_set_rxnfc(struct net_device *dev, struct ethtool_rxnfc *nfc)
 			ret = otx2_add_flow(pfvf, &nfc->fs);
 		break;
 	case ETHTOOL_SRXCLSRLDEL:
-		ret = otx2_remove_flow(pfvf, nfc->fs.location);
+		if (netif_running(dev) && ntuple)
+			ret = otx2_remove_flow(pfvf, nfc->fs.location);
 		break;
 	default:
 		break;
diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
index b1e5bcc7981d..a6e99ac62682 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
@@ -1685,6 +1685,7 @@ int otx2_stop(struct net_device *netdev)
 	/* First stop packet Rx/Tx */
 	otx2_rxtx_enable(pf, false);
 
+	otx2_destroy_ethtool_flows(pf);
 
 	pf->intf_down = true;
 	/* 'intf_down' may be checked on any cpu */
-- 
2.17.1

